Bugsnag/Breadcrumbs/BSGNetworkBreadcrumb.m

```diff
@@ -0,0 +1,96 @@
+//
+//  BSGNetworkBreadcrumb.m
+//  Bugsnag
+//
+//  Created by Nick Dowell on 24/08/2022.
+//  Copyright © 2022 Bugsnag Inc. All rights reserved.
+//
+
+#import "BSGNetworkBreadcrumb.h"
+
+BugsnagBreadcrumb * BSGNetworkBreadcrumbWithTaskMetrics(NSURLSessionTask *task, NSURLSessionTaskMetrics *metrics) {
+    NSURLRequest *request = task.originalRequest ? task.originalRequest : task.currentRequest;
+    if (!request) {
+        return nil;
+    }
+
+    NSMutableDictionary *metadata = [NSMutableDictionary dictionary];
+    metadata[@"duration"] = @((unsigned)(metrics.taskInterval.duration * 1000));
+    metadata[@"method"] = request.HTTPMethod;
+
+    NSURL *url = request.URL;
+    if (url) {
+        NSURLComponents *components = [NSURLComponents componentsWithURL:url resolvingAgainstBaseURL:YES];
+        metadata[@"url"] = BSGURLStringForComponents(components);
+        metadata[@"urlParams"] = BSGURLParamsForQueryItems(components.queryItems);
+    }
+
+    if (task.countOfBytesSent) {
+        metadata[@"requestContentLength"] = @(task.countOfBytesSent);
+    } else if (request.HTTPBody) {
+        // Fall back because task.countOfBytesSent is 0 when a custom NSURLProtocol is used
+        metadata[@"requestContentLength"] = @(request.HTTPBody.length);
+    }
+
+    NSString *message = @"NSURLSession request error";
+
+    // NSURLSessionTaskTransactionMetrics.response is nil when a custom NSURLProtocol is used. 
+    // If there was an error, task.response will be nil.
+    NSURLResponse *response = task.response; 
+    if ([response isKindOfClass:[NSHTTPURLResponse class]]) {
+        NSInteger statusCode = ((NSHTTPURLResponse *)response).statusCode;
+        if (100 <= statusCode && statusCode < 400) {
+            message = @"NSURLSession request succeeded";
+        }
+        if (400 <= statusCode && statusCode < 500) {
+            message = @"NSURLSession request failed";
```

Comment:
Do we put any message when we get an error response (500)?

Reply:
Yes - `message` is initialized with "NSURLSession request error" on line 35.